03_DeviceSelectionAndSharedSources/main.cpp

```diff
@@ -13,318 +13,422 @@ using namespace system;
-				options.spirvOptimizer = nullptr;
-				options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_SOURCE_BIT;
-				// The nice thing is that when you load a shader from file, it has a correctly set `filePathHint`
-				// so it plays nicely with the preprocessor, and finds `#include`s without intervention.
-				options.preprocessorOptions.sourceIdentifier = unspecialized->getFilepathHint();
-				options.preprocessorOptions.logger = m_logger.get();
-				options.preprocessorOptions.includeFinder = compilerSet->getShaderCompiler(unspecialized->getContentType())->getDefaultIncludeFinder();
-
-				auto spirvUnspecialized = compilerSet->compileToSPIRV(unspecialized,options);
-				const CSPIRVIntrospector::SIntrospectionParams inspctParams = {.entryPoint=source->getSpecializationInfo().entryPoint,.cpuShader=spirvUnspecialized};
-
-				introspection = introspector->introspect(inspctParams);
-				if (!introspection)
-					return logFail("SPIR-V Introspection failed, probably the required SPIR-V compilation failed first!");
-
-				// now we need to swap out the HLSL for SPIR-V
-				source = make_smart_refctd_ptr<ICPUSpecializedShader>(std::move(spirvUnspecialized),ISpecializedShader::SInfo(nullptr,nullptr,source->getSpecializationInfo().entryPoint));
-			}
+			m_logger->log("------- test.hlsl INTROSPECTION -------", ILogger::E_LOG_LEVEL::ELL_WARNING);
+			CSPIRVIntrospector introspector;
+			auto sourceIntrospectionPair = this->compileShaderAndTestIntrospection("app_resources/test.hlsl", introspector);
+			auto pplnIntroData = core::make_smart_refctd_ptr<CSPIRVIntrospector::CPipelineIntrospectionData>();
+			pplnIntroData->merge(sourceIntrospectionPair.second.get());
+		}
+#endif
+		// testing creation of compute pipeline layouts compatible for multiple shaders
+		{
+			constexpr std::array mergeTestShadersPaths = {
+				"app_resources/pplnLayoutMergeTest/shader_1.comp.hlsl",
+				"app_resources/pplnLayoutMergeTest/shader_2.comp.hlsl",
+				"app_resources/pplnLayoutMergeTest/shader_3.comp.hlsl",
+				"app_resources/pplnLayoutMergeTest/shader_4.comp.hlsl"
+			};
+			constexpr uint32_t MERGE_TEST_SHADERS_CNT = mergeTestShadersPaths.size();
 
-			// Just a check that out specialization info will match
-			if (!introspection->canSpecializationlesslyCreateDescSetFrom())
-				return logFail("Someone changed the shader and some descriptor binding depends on a specialization constant!");
+			auto confirmExpectedOutput = [this](bool value, bool expectedValue)
+				{
+					if (value != expectedValue)
+					{
+						m_logger->log("\"CSPIRVIntrospector::CPipelineIntrospectionData::merge\" function FAIL, incorrect output.",
+							ILogger::E_LOG_LEVEL::ELL_ERROR);
+					}
+					else
+					{
+						m_logger->log("\"CSPIRVIntrospector::CPipelineIntrospectionData::merge\" function SUCCESS, correct output.",
+							ILogger::E_LOG_LEVEL::ELL_PERFORMANCE);
+					}
```

Comment:
imo you should do the merge of stage to pipeline in the lambda and debug print somewhere the pipeline layout introspection post-merge

Reply:
that will require additional `debugPrint` function in `CPipelineIntrospectionData`